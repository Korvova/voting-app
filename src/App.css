/* Существующие стили */
.app {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  height: 100vh;
  background-color: #f0f2f5;
  margin: 0;
  padding: 0;
  width: 100%;
  box-sizing: border-box;
}

.title {
  font-size: 3rem;
  font-weight: bold;
  color: #1a73e8;
  margin-bottom: 2rem;
  text-align: center;
}

.login-form {
  background: white;
  padding: 2rem;
  border-radius: 8px;
  box-shadow: 0 2px 8px rgba(0, 0, 0, 0.1);
  width: 100%;
  max-width: 400px;
}

.form-group {
  margin-bottom: 1.5rem;
}

.form-group label {
  display: block;
  margin-bottom: 0.5rem;
  font-weight: 500;
  color: #333;
}

.form-group input {
  width: 100%;
  padding: 0.75rem;
  border: 1px solid #ccc;
  border-radius: 4px;
  font-size: 1rem;
}

.form-group input:focus {
  outline: none;
  border-color: #1a73e8;
  box-shadow: 0 0 0 2px rgba(26, 115, 232, 0.2);
}

.form-group input[type="datetime-local"] {
  width: 200px; /* Уменьшаем размер инпутов для даты/времени */
}

button {
  width: 100%;
  padding: 0.75rem;
  background-color: #1a73e8;
  color: white;
  border: none;
  border-radius: 4px;
  font-size: 1rem;
  cursor: pointer;
  margin-top: 1rem;
}

button:hover {
  background-color: #1557b0;
}

.forgot-password {
  color: #1a73e8;
  text-align: center;
  cursor: pointer;
  margin-top: 1rem;
}

.forgot-password:hover {
  text-decoration: underline;
}

.modal {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background: rgba(0, 0, 0, 0.5);
  display: flex;
  align-items: center;
  justify-content: center;
}

.modal-content {
  background: white;
  padding: 2rem;
  border-radius: 8px;
  text-align: center;
  width: 100%;
  max-width: 800px; /* Увеличиваем ширину */
  max-height: 80vh; /* Ограничиваем высоту */
  overflow-y: auto; /* Добавляем скролл */
}

.modal-content p {
  margin-bottom: 1.5rem;
}

.modal-content button {
  width: auto;
  padding: 0.5rem 1rem;
  margin: 0 0.5rem;
}

.admin-panel {
  display: flex;
  flex-direction: column;
  min-height: 100vh;
  background-color: #f4f6f9;
}

.header {
  background-color: #1a73e8;
  color: white;
  padding: 1rem;
  display: flex;
  justify-content: flex-end;
  align-items: center;
}

.user-info {
  display: flex;
  align-items: center;
  gap: 1rem;
}

.logout-button {
  background-color: #d93025;
  padding: 0.5rem 1rem;
  font-size: 0.9rem;
  margin-top: 0;
}

.logout-button:hover {
  background-color: #b71c1c;
}

.menu {
  display: flex;
  gap: 1rem;
  padding: 1rem;
  background-color: #e8eaf6;
  justify-content: center;
}

.menu button {
  background-color: #3949ab;
  padding: 0.75rem 1.5rem;
  font-size: 1rem;
  border-radius: 6px;
  transition: background-color 0.3s;
}

.menu button:hover {
  background-color: #1a237e;
}

.menu button.active {
  background-color: #1a237e;
  box-shadow: 0 0 8px rgba(0, 0, 0, 0.2);
}

.content {
  flex: 1;
  padding: 0.1rem;
  text-align: center;
}

.users-page {
  padding: 1rem;
}

.add-button {
  background-color: #34c759;
  padding: 0.75rem 1.5rem;
  margin-bottom: 1rem;
  font-size: 1rem;
  border-radius: 6px;
  transition: background-color 0.3s;
}

.add-button:hover {
  background-color: #2ea44f;
}

.users-table {
  width: 100%;
  border-collapse: collapse;
  background: white;
  box-shadow: 0 2px 8px rgba(0, 0, 0, 0.1);
  border-radius: 8px;
  overflow: hidden;
   margin-bottom: 20px;
}

.users-table th,
.users-table td {
  padding: 1rem;
  text-align: left;
  border-bottom: 1px solid #e0e0e0;
}

.users-table th {
  background-color: #f5f5f5;
  font-weight: 600;
  color: #333;
}

.users-table td {
  color: #555;
}

.users-table tr:hover {
  background-color: #f9f9f9;
}

.delete-button {
  background-color: #d93025;
  padding: 0.25rem 0.5rem;
  font-size: 0.9rem;
  margin-top: 0;
  width: auto;
}

.delete-button:hover {
  background-color: #b71c1c;
}

.divisions-page {
  padding: 2rem;
}

.divisions-table {
  width: 100%;
  border-collapse: collapse;
  background: white;
  box-shadow: 0 2px 8px rgba(0, 0, 0, 0.1);
  border-radius: 8px;
  overflow: hidden;
}

.divisions-table th,
.divisions-table td {
  padding: 1rem;
  text-align: left;
  border-bottom: 1px solid #e0e0e0;
}

.divisions-table th {
  background-color: #f5f5f5;
  font-weight: 600;
  color: #333;
}

.divisions-table td {
  color: #555;
}

.divisions-table tr:hover {
  background-color: #f9f9f9;
}

.meetings-page {
  padding: 2rem;
}

.meetings-header {
  display: flex;
  align-items: center;
  gap: 1rem;
  margin-bottom: 1rem;
}

.archive-button {
  background-color: #757575;
  padding: 0.75rem 1.5rem;
  font-size: 1rem;
  border-radius: 6px;
  transition: background-color 0.3s;
}

.archive-link {
  color: #1a73e8;
  text-decoration: none;
  font-size: 1rem;
}

.archive-link:hover {
  text-decoration: underline;
}

.archive-button:hover {
  background-color: #616161;
}

.meetings-table {
  width: 100%;
  border-collapse: collapse;
  background: white;
  box-shadow: 0 2px 8px rgba(0, 0, 0, 0.1);
  border-radius: 8px;
  overflow: hidden;
}

.meetings-table th,
.meetings-table td {
  padding: 1rem;
  text-align: left;
  border-bottom: 1px solid #e0e0e0;
}

.meetings-table th {
  background-color: #f5f5f5;
  font-weight: 600;
  color: #333;
}

.meetings-table td {
  color: #555;
}

.meetings-table:not(.control-meetings-table) tr:hover {
  background-color: #f9f9f9;
}

.meetings-table a {
  color: #1a73e8;
  text-decoration: none;
}

.meetings-table a:hover {
  text-decoration: underline;
}

.meetings-table .divisions-column {
  white-space: pre-line; /* Позволяет переносить текст на новую строку */
}

.archive-icon {
  background: none;
  border: none;
  cursor: pointer;
  font-size: 1.2rem;
}

.divisions-list {
  margin: 1rem 0;
}

.division-item {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 0.5 rem;
  border-bottom: 1px solid #e0e0e0;
}

.division-item button {
  background: none;
  border: none;
  cursor: pointer;
  font-size: 1.2rem;
  margin-top: 0;
}

.agenda-table {
  width: 100%;
  border-collapse: collapse;
  margin: 1rem 0;
}

.agenda-table th,
.agenda-table td {
  padding: 0.5rem;
  border: 1px solid #e0e0e0;
}

.agenda-table th {
  background-color: #f5f5f5;
  font-weight: 600;
}

.agenda-table input,
.agenda-table select {
  width: 100%;
  padding: 0.5rem;
  border: 1px solid #ccc;
  border-radius: 4px;
}

.agenda-table input,
.agenda-table select {
  width: 100%;
  max-width: 100%; /* Ограничиваем ширину */
  box-sizing: border-box; /* Учитываем padding и border в ширине */
}

.add-agenda-button {
  background-color: #34c759;
  padding: 0.5rem 1rem;
  font-size: 0.9rem;
  margin: 1rem 0;
  border-radius: 6px;
}

.add-agenda-button:hover {
  background-color: #2ea44f;
}

.modal-buttons {
  display: flex;
  flex-direction: column;
  gap: 1rem;
  margin-top: 2rem; /* Увеличиваем отступ сверху для большего пространства */
}

.action-buttons {
  display: flex;
  justify-content: center;
  gap: 2rem;
}

.control-page {
  padding: 2rem;
}

.control-page h2 {
  margin-bottom: 1rem;
  font-size: 1.5rem;
  color: #333;
}

.control-table {
  width: 100%;
  border-collapse: collapse;
  background: white;
  box-shadow: 0 2px 8px rgba(0, 0, 0, 0.1);
  border-radius: 8px;
  overflow: hidden;
}

.control-table th,
.control-table td {
  padding: 1rem;
  text-align: left;
  border-bottom: 1px solid #e0e0e0;
}

.control-table th {
  background-color: #f5f5f5;
  font-weight: 600;
  color: #333;
}

.control-table td {
  color: #555;
}

.control-table tr:hover {
  background-color: #f9f9f9;
}

.vote-results-column {
  white-space: pre-line; /* Отображение итогов в столбик */
}

.voting-active {
  background-color: #e0ffe0; /* Светло-зелёный фон для активного голосования */
}

.voting-completed {
  background-color: #e0e0e0; /* Серый фон для завершённого голосования */
}

.vote-results .action-buttons {
  display: flex;
  justify-content: center;
  gap: 1rem;
  margin-top: 1rem;
}

.vote-results .action-buttons button {
  width: 100px; /* Уменьшаем ширину кнопок "Применить" и "Отмена" */
}

.start-meeting-button {
  width: 180px; /* Уменьшаем ширину кнопки "Начать заседание" */
}

.start-meeting-button.start {
  background-color: #34c759; /* Зелёный цвет для "▶ Начать заседание" */
}

.start-meeting-button.start:hover {
  background-color: #2ea44f;
}

.start-meeting-button.end {
  background-color: #f40e0e; /* Красный цвет для "⏹ Завершить заседание" */
}

.start-meeting-button.end:hover {
  background-color: #dd1e1e;
}

.start-meeting-button.completed {
  background-color: #a28663; /* Оранжево-серый цвет для "✔ Завершено" */
  color: #666;
}

.action-column {
  display: flex;
  align-items: center;
  gap: 0.5rem;
}

.vote-button,
.end-vote-button {
  width: 80px; /* Одинаковая ширина для кнопок "▷" и "Завершить" */
  padding: 0.5rem;
  font-size: 0.9rem;
  border-radius: 4px;
  display: flex;
  align-items: center;
  justify-content: center;
}

.vote-button {
  background-color: #1a73e8;
  color: white;
}

.vote-button:disabled {
  background-color: #b0b0b0; /* Серый цвет для неактивной кнопки "▷" */
  cursor: not-allowed;
}

.end-vote-button {
  background-color: #ff5733; /* Оранжево-красный цвет для кнопки "Завершить" */
  color: white;
}

.end-vote-button:hover {
  background-color: #e64a19;
}

.completed-text {
  color: #888;
  font-size: 0.9rem;
  margin-left: 0.5rem;
}

.start-warning-button {
  background-color: #34c759; /* Зелёный цвет для кнопки "Запустить" */
}

.start-warning-button:hover {
  background-color: #2ea44f;
}

.user-page {
  display: flex;
  flex-direction: column;
  min-height: 100vh;
  background-color: #f4f6f9;
}

.meeting-container {
  position: relative;
  max-width: 1200px;
  margin: 0 auto;
  padding: 2rem;
  display: grid;
  grid-template-areas: 
    "agenda participants";
  grid-template-columns: 65% 30%;
  gap: 2rem;
}

.agenda-section {
  grid-area: agenda;
  width: 100%;
}

.agenda-items-table {
  width: 100%; /* Таблица занимает всю ширину своего контейнера */
  border-collapse: collapse;
  background: #f9f9f9;
  border-radius: 4px;
  overflow: hidden;
}

.agenda-items-table th,
.agenda-items-table td {
  padding: 0.75rem;
  border-bottom: 1px solid #e0e0e0;
}

.agenda-items-table th {
  background-color: #f5f5f5;
  font-weight: 600;
  color: #333;
}

.agenda-items-table td {
  color: #555;
}

.agenda-items-table tr:hover {
  background-color: #f0f0f0;
}

.agenda-items-table .active-agenda-item {
  background-color: #e0ffe0; /* Светло-зелёный цвет для текущего вопроса */
}

.participants-section {
  grid-area: participants;
  position: sticky;
  top: 2rem;
  width: 100%;
}

.participants-block {
  padding: 1rem;
  background: #e8eaf6;
  border-radius: 4px;
  text-align: left;
}

.participants-block p {
  font-size: 1.1rem;
  font-weight: 600;
  color: #333;
  margin-bottom: 0.5rem;
}

.participants-block h3 {
  font-size: 1.2rem;
  color: #333;
  margin-bottom: 0.5rem;
}

.participants-list {
  max-height: 200px; /* Высота для отображения примерно 10 участников */
  overflow-y: auto; /* Добавляем скролл */
  list-style: none;
  padding: 0;
  margin: 0;
}

.participants-list li {
  font-size: 1rem;
  margin: 0.3rem 0;
  color: #555;
}

.protocol-link {
  display: inline-block;
  margin-top: 1rem;
  color: #1a73e8;
  text-decoration: none;
  font-size: 1rem;
}

.protocol-link:hover {
  text-decoration: underline;
}

.protocol-page {
  padding: 2rem;
}

.protocol-page h2 {
  margin-bottom: 1rem;
  font-size: 1.5rem;
  color: #333;
}

.protocol-table {
  width: 100%;
  border-collapse: collapse;
  background: white;
  box-shadow: 0 2px 8px rgba(0, 0, 0, 0.1);
  border-radius: 8px;
  overflow: hidden;
}

.protocol-table th,
.protocol-table td {
  padding: 1rem;
  text-align: left;
  border-bottom: 1px solid #e0e0e0;
}

.protocol-table th {
  background-color: #f5f5f5;
  font-weight: 600;
  color: #333;
}

.protocol-table td {
  color: #555;
}

.protocol-table tr:hover {
  background-color: #f9f9f9;
}

.vote-buttons {
  display: flex;
  justify-content: center;
  gap: 1rem;
  margin-top: 1rem;
}

.vote-button {
  width: 100px;
  padding: 0.75rem;
  font-size: 1rem;
  border-radius: 4px;
  border: none;
  cursor: pointer;
  background-color: #1a73e8;
  color: white;
}

.vote-button:hover:not(:disabled) {
  background-color: #1557b0;
}

.vote-button:disabled {
  background-color: #b0b0b0;
  cursor: not-allowed;
}

.vote-button-selected {
  width: 100px;
  padding: 0.75rem;
  font-size: 1rem;
  border-radius: 4px;
  border: none;
  cursor: pointer;
  background-color: #34c759;
  color: white;
}

/* Стили для строк таблицы в зависимости от статуса заседания */
.completed-meeting {
  background-color: #ffcc99; /* Оранжево-серая заливка для COMPLETED */
  color: #666; /* Серый текст */
}

.in-progress-meeting {
  background-color: #ccffcc; /* Зелёная заливка для IN_PROGRESS */
  color: #333; /* Тёмный текст для читаемости */
}

.agenda-table .active-agenda-item {
  background-color: #e0ffe0 !important; /* Добавляем !important для теста */
}


/* Стили для иконки статуса */
.status-icon {
  cursor: pointer;
  font-size: 1.2rem;
  font-weight: bold;
}

.status-icon.online {
  color: #34c759; /* Зеленый для онлайн */
}

.status-icon.offline {
  color: #d93025; /* Красный для оффлайн */
}




.modal-buttons button:first-child {
  background-color: #25d95b; /* Красный для "Да" */
  color: white;
}

.modal-buttons button:last-child {
  background-color: #34c759; /* Зеленый для "Отмена" */
  color: white;
}

/* Стили для страницы конфигурации */
.config-page {
  padding: 2rem;
}

.config-menu {
  display: flex;
  gap: 1rem;
  padding: 1rem;
  background-color: #e8eaf6;
  justify-content: center;
  margin-bottom: 1rem;
}

.config-menu a {
  background-color: #3949ab;
  padding: 0.75rem 1.5rem;
  font-size: 1rem;
  border-radius: 6px;
  transition: background-color 0.3s;
  color: white;
  text-decoration: none;
}

.config-menu a:hover {
  background-color: #1a237e;
}

.config-menu a.active {
  background-color: #1a237e;
  box-shadow: 0 0 8px rgba(0, 0, 0, 0.2);
}

.config-subpage {
  background: white;
  padding: 2rem;
  border-radius: 8px;
  box-shadow: 0 2px 8px rgba(0, 0, 0, 0.1);
  text-align: center;
}


/* Стили для таблицы процедур подсчета голосов */
.procedures-table {
  width: 100%;
  border-collapse: collapse;
  background: white;
  box-shadow: 0 2px 8px rgba(0, 0, 0, 0.1);
  border-radius: 8px;
  overflow: hidden;
  margin-bottom: 20px;
}

.procedures-table th,
.procedures-table td {
  padding: 1rem;
  text-align: left;
  border-bottom: 1px solid #e0e0e0;
}

.procedures-table th {
  background-color: #f5f5f5;
  font-weight: 600;
  color: #333;
}

.procedures-table td {
  color: #555;
}

.procedures-table tr:hover {
  background-color: #f9f9f9;
}


/* Стили для конструктора условий */
.condition-block {
  margin-bottom: 1rem;
  padding: 1rem;
  background-color: #f9f9f9;
  border-radius: 4px;
}

.condition-elements {
  display: flex;
  flex-wrap: wrap;
  gap: 0.5rem;
  margin-bottom: 0.5rem;
}

.condition-element {
  display: flex;
  align-items: center;
  gap: 0.5rem;
}

.condition-element select,
.condition-element input {
  padding: 0.5rem;
  border: 1px solid #ccc;
  border-radius: 4px;
}

.condition-element button {
  width: auto;
  padding: 0.5rem;
  background-color: #d93025;
  color: white;
  border-radius: 4px;
}

.condition-element button:hover {
  background-color: #b71c1c;
}


/* Стили для кнопок в колонке "Действия" */
.actions-column {
  display: flex;
  gap: 0.5rem;
  align-items: center;
}

.actions-column button {
  width: 100px; /* Фиксированная ширина для одинакового размера */
  padding: 0.5rem;
  font-size: 0.9rem;
  border-radius: 4px;
  border: none;
  cursor: pointer;
}

.actions-column button:first-child {
  background-color: #1a73e8; /* Синий для "Редактировать" */
  color: white;
}

.actions-column button:first-child:hover {
  background-color: #1557b0;
}

.actions-column button:last-child {
  background-color: #d93025; /* Красный для "Удалить" */
  color: white;
}

.actions-column button:last-child:hover {
  background-color: #b71c1c;
}
