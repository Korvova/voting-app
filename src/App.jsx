import { BrowserRouter as Router, Route, Routes, Navigate } from 'react-router-dom';
import LoginPage from './pages/LoginPage';
import AdminPanel from './pages/AdminPanel';
import UserPage from './pages/UserPage';
import ProtocolPage from './pages/ProtocolPage';
import './App.css';
import { useState } from 'react';
import axios from 'axios';

function App() {
  const [user, setUser] = useState(null);

  const handleLogin = async (userData) => {
    try {
      const response = await axios.post('http://217.114.10.226:5000/api/login', userData);
      if (response.data.success) {
        setUser(response.data.user);
      } else {
        console.log('Login failed:', response.data.error);
      }
    } catch (error) {
      console.log('Login error:', error.message);
    }
  };

  const handleLogout = () => {
    setUser(null);
  };

  return (
    <Router>
      <Routes>
        <Route
          path="/"
          element={
            user ? (
              user.isAdmin ? (
                <Navigate to="/admin" />
              ) : (
                <Navigate to="/user" />
              )
            ) : (
              <LoginPage onLogin={handleLogin} />
            )
          }
        />
        <Route
          path="/admin/*"
          element={
            user && user.isAdmin ? (
              <AdminPanel user={user} onLogout={handleLogout} />
            ) : (
              <Navigate to="/" />
            )
          }
        />
        <Route
          path="/user/*"
          element={
            user && !user.isAdmin ? (
              <UserPage user={user} onLogout={handleLogout} />
            ) : (
              <Navigate to="/" />
            )
          }
        >
          <Route path="protocol/:id" element={<ProtocolPage />} />
        </Route>
      </Routes>
    </Router>
  );
}

export default App;