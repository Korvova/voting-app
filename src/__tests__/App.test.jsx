import React from 'react';
import { render, screen, waitFor, act } from '@testing-library/react';
import '@testing-library/jest-dom';
import io from 'socket.io-client';
import App from '../App';

jest.mock('socket.io-client', () => {
  const mockSocket = {
    on: jest.fn(),
    emit: jest.fn(),
    disconnect: jest.fn(),
    connected: true,
  };
  return jest.fn(() => mockSocket);
});

describe('App Component', () => {
  test('показывает модальный режим при получении события, изменившего голосование', async () => {
    const mockSocket = io();
    render(<App />);

    await act(async () => {
      const voteChangedHandler = mockSocket.on.mock.calls.find(call => call[0] === 'vote-changed')[1];
      voteChangedHandler({ optionId: 1, votes: 8 });
    });

    await waitFor(() => {
      expect(screen.getByText('Vote Updated')).toBeInTheDocument();
    }, { timeout: 1000 });

    mockSocket.disconnect();
    expect(mockSocket.disconnect).toHaveBeenCalled();
  });
});