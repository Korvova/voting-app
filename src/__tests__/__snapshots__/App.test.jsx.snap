// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App UI сохраняет snapshot HTML главной страницы 1`] = `
"<!DOCTYPE html><html lang="en"><head>
    <meta charset="UTF-8">
    <link rel="icon" type="image/svg+xml" href="/vite.svg">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Vite + React</title>
    <script type="module" crossorigin="" src="/assets/index-CrLHBHvD.js"></script>
    <link rel="stylesheet" crossorigin="" href="/assets/index-DyMikYCs.css">
  </head>
  <body>
    <div id="root"><div class="container"><h1 class="title">RMS - Система голосования</h1><div class="login-form"><input placeholder="Электронная почта" class="input" type="email"><input placeholder="Пароль" class="input" type="password"><a href="#" class="forgot-password">Напомнить пароль</a><button class="login-button">Войти</button></div></div></div>
  

</body></html>"
`;
